### PR TITLE
feat(cve): multi-stage NVD search with version-based filtering

### LINKED ISSUE #105: Enhancement: CVE Lookup Accuracy by Adding Multi-Stage NVD Search and Version-Based Filtering
## Description

The current `cve_lookup` implementation performs a single NVD API query using:

```python
keywordSearch = f"{software} {version}"
```

This approach is often inaccurate because the NVD API's `keywordSearch` parameter performs text-based matching. Many CVE records do not explicitly mention the exact software version in their descriptions, causing valid vulnerabilities to be missed.

For example:

* `nginx 1.24.0` may return zero results even though vulnerabilities affecting that version exist.
* CVE records frequently describe affected versions using ranges such as:

  * `versions below 1.30.1`
  * `versions 1.19.0 through 1.30.0`
* The exact string `"nginx 1.24.0"` is often absent from the CVE description.

As a result, the current implementation can produce false negatives.

## Proposed Solution

Implement a multi-stage CVE lookup strategy using the official NVD API.

### Stage 1: Search Using Software Name and Version

Perform the existing lookup:

```python
keywordSearch = f"{software} {version}"
```

If vulnerabilities are returned, process and return them as usual.

### Stage 2: Fallback to Software Name Only

If no vulnerabilities are found, perform a second lookup using only the software name:

```python
keywordSearch = software
```

This typically returns all CVEs associated with the software.

### Stage 3: Filter Vulnerabilities by Version

For each returned CVE:

* Parse the `configurations` field.
* Extract version constraints from the `cpeMatch` objects, such as:

  * `versionStartIncluding`
  * `versionStartExcluding`
  * `versionEndIncluding`
  * `versionEndExcluding`

Compare the supplied software version against these constraints to determine whether the vulnerability affects the target version.

Version comparisons can be performed using Python's `packaging.version.Version`.

Only vulnerabilities that match the supplied version should be returned.

## Proposed Workflow

```text
Input:
Software = nginx
Version = 1.24.0

↓

Search:
"nginx 1.24.0"

↓

Results Found?

YES
↓
Return results

NO
↓
Search:
"nginx"

↓

Retrieve all nginx CVEs

↓

Parse version constraints from NVD configurations

↓

Compare:
1.24.0

↓

Return only matching vulnerabilities
```

## Benefits

* Significantly improves CVE lookup accuracy.
* Reduces false negatives when exact version strings are not present in CVE descriptions.
* Uses only the official NVD API.
* Requires no API keys or third-party services.
* Avoids maintaining software-to-CPE mapping databases.
* Provides version-aware vulnerability matching while keeping the implementation relatively simple.

## Expected Outcome

The `cve_lookup` tool will be able to detect vulnerabilities affecting a specific software version even when the exact version is not explicitly mentioned in the NVD keyword search results, resulting in more accurate and reliable vulnerability assessments.

Files to touch :- tools/cve_tool.py
tests/test_cve_lookup.py

### PR DESCRIPTION
**Pull request by 01luyicheng**
# CVE Lookup: Multi-Stage NVD Search with Version-Based Filtering

Closes #105.

## Summary

Replaces the single-query `cve_lookup` implementation with a 3-stage strategy that significantly reduces false negatives. The previous tool searched NVD with `keywordSearch = f"{software} {version}"`, which is text-based — many CVEs don't mention the exact version string in their descriptions, so legitimate vulnerabilities were silently missed. This PR adds a fallback search (Stage 2) and CPE version-range filtering (Stage 3) using the official NVD API only — no new API keys, no third-party services.

## Why This Change

- **Fixes false negatives**: A security-recon tool that silently misses real vulnerabilities (e.g., `nginx 1.24.0` returning 0 results when CVEs affect that version) undermines the tool's core purpose.
- **Uses only the official NVD API**: No API keys, no third-party CPE-mapping databases to maintain.
- **Maintainer-requested**: Issue #105 is labeled `help wanted` + `size: medium`.
- **Backward-compatible**: Existing return shape (`success`/`software`/`version`/`total_results`/`cves`) is preserved. A new `version_filtering_applied` boolean is added.

## The 3-Stage Strategy

```
Stage 1: keywordSearch = "{software} {version}"
   │
   ├── Results found? → return them (version_filtering_applied: false)
   │
   └── No results ↓

Stage 2: keywordSearch = "{software}"  (broader query)
   │
   ↓
Stage 3: For each CVE returned, parse configurations[].cpeMatch[] for
         versionStartIncluding / versionStartExcluding /
         versionEndIncluding / versionEndExcluding.
         Use packaging.version.Version to compare the supplied version
         against each cpeMatch range. Keep only CVEs whose range includes
         the target version.
   │
   ↓
Return filtered CVEs (version_filtering_applied: true)
```

### `tools/cve_tool.py`
- Added `from packaging.version import Version, InvalidVersion` import.
- Refactored into helpers (DRY):
  - `_query_nvd(keyword: str) -> list[dict]` — wraps the HTTP call; returns the raw `vulnerabilities` list.
  - `_simplify_cve(item: dict) -> dict` — converts a raw NVD item to the existing simplified shape (`cve_id`, `severity`, `score`, `published`, `last_modified`, `description`).
  - `_cve_affects_version(cve: dict, target_version: str) -> bool` — parses `configurations[].nodes[].cpeMatch[]` for version ranges and returns `True` if any vulnerable cpeMatch range includes the target version.
  - `_version_in_range(target: Version, match: dict) -> bool` — applies the four boundary constraints (`versionStartIncluding`/`Excluding`, `versionEndIncluding`/`Excluding`); defensively handles `InvalidVersion` by skipping that cpeMatch.
- Main `cve_lookup` flow:
  - Stage 1 — query with `"{software} {version}"`. If non-empty, return existing shape with `version_filtering_applied: False`.
  - Stage 2 — query with `"{software}"`.
  - Stage 3 — filter the Stage 2 results with `_cve_affects_version`. Return the filtered list with `version_filtering_applied: True`.
- **Preserved**:
  - Function signature `def cve_lookup(software: str, version: str) -> dict:`.
  - Input validation (`.strip()` + required check).
  - All 4 existing exception handlers (`HTTPError`, `Timeout`, `RequestException`, generic `Exception`).
  - Existing CVE dict shape (`cve_id`, `severity`, `score`, `published`, `last_modified`, `description`).
- **Edge case handling**:
  - If target version is unparseable by `packaging.version.Version` (e.g., `"abc"`), the filter is bypassed and Stage 2 results are returned as-is (fail-open — better to return too many CVEs than to silently drop them all).
  - cpeMatch entries with `vulnerable: False` are skipped (only vulnerable CPE matches count).
  - cpeMatch entries with unparseable constraint versions are skipped (one bad constraint doesn't fail the whole CVE).
  - CVEs with no `configurations` field are excluded from the filtered results (no version info → can't confirm it affects the target).

### `tests/test_cve_lookup.py`
- **All 7 existing tests preserved and passing** (no behavior change for Stage 1 hit path or error paths).
- **13 new tests added**:
  1. `test_stage1_returns_results_without_filtering` — Stage 1 hit → `version_filtering_applied=False`, only 1 NVD call made.
  2. `test_stage2_stage3_falls_back_and_filters` — Stage 1 empty, Stage 2 returns 3 CVEs, filter keeps 2 → `version_filtering_applied=True`, 2 NVD calls made.
  3. `test_stage3_filters_out_non_matching_version` — both Stage 2 CVEs out of range → 0 CVEs returned.
  4. `test_cve_with_no_configurations_excluded_by_filter` — CVE with no `configurations` key → excluded.
  5. `test_invalid_target_version_skips_filter` — target `"abc"` → Stage 2 results returned as-is (filter skipped, fail-open).
  6. `test_version_start_including` — `versionStartIncluding: "1.0.0"` boundary: `1.0.0` included, `0.9.9` excluded.
  7. `test_version_start_excluding` — `versionStartExcluding: "1.0.0"` boundary: `1.0.0` excluded, `1.0.1` included.
  8. `test_version_end_including` — `versionEndIncluding: "2.0.0"` boundary: `2.0.0` included, `2.0.1` excluded.
  9. `test_version_end_excluding` — `versionEndExcluding: "2.0.0"` boundary: `1.9.9` included, `2.0.0` excluded.
  10. `test_non_vulnerable_cpe_match_ignored` — cpeMatch with `vulnerable: False` is ignored.
  11. `test_cpe_match_with_no_constraints_matches_any_version` — a vulnerable cpeMatch with no version constraints matches any target.
  12. `test_invalid_constraint_version_skips_match` — unparseable constraint version (e.g. `"*"`) skips the cpeMatch defensively.
  13. `test_stage1_http_error_does_not_fall_through_to_stage2` — Stage 1 HTTPError returns the error and does NOT fire Stage 2.
- The pre-existing `test_cve_lookup_empty_results` was strengthened to assert `version_filtering_applied=True` and exactly 2 NVD calls.
- Tests use `mock_get.side_effect` to simulate Stage 1 → Stage 2 sequential NVD responses.
- No real network calls — all `requests.get` calls are mocked.

### `requirements.txt` / `pyproject.toml` / `uv.lock`
- Added `packaging==26.2` to `requirements.txt` (matches uv.lock resolution).
- Added `packaging>=26.2` to `pyproject.toml` dependencies (kept alphabetical order).
- Regenerated `uv.lock` so `packaging` is listed as a direct dependency (matches the precedent set by the tldextract PR `6d611d9`).

### What's NOT changed (intentionally)
- **`server.py`** — Import and `mcp.tool()` registration unchanged.
- **`mcp.json`** — Tool metadata unchanged; return shape preserved.
- **`README.md`** — Description unchanged; tool still "no API key required".
- **`utils/helpers.py`** — `get_cvss_details` and `get_english_description` reused unchanged.

## Verification

```
$ python -m pytest tests/test_cve_lookup.py -v
20 passed

$ python -c "from tools.cve_tool import cve_lookup; print(cve_lookup('apache', ''))"
{'success': False, 'error': 'Software and version are required'}

$ python -c "from tools.cve_tool import _cve_affects_version; print(_cve_affects_version({'configurations':[{'nodes':[{'cpeMatch':[{'vulnerable': True, 'versionStartIncluding': '1.0.0', 'versionEndIncluding': '2.0.0'}]}]}]}, '1.5.0'))"
True
```

## Files Changed

```
 pyproject.toml           |   1 +
 requirements.txt         |   1 +
 tests/test_cve_lookup.py | 332 ++++++++++++++++++++++++++++++++++++++++++++++-
 tools/cve_tool.py        | 130 +++++++++++++++----
 uv.lock                  |   2 +
 5 files changed, 439 insertions(+), 27 deletions(-)
```

## Checklist

- [x] Tool follows existing `try/except` + `{"success": ...}` pattern from `contributing.md`.
- [x] Input validation preserved (required software + version).
- [x] Existing return shape preserved (no breaking change to downstream consumers).
- [x] All 7 existing tests preserved and passing.
- [x] 13 new tests added covering all 3 stages and version filtering edge cases (including all 4 boundary types: Start/End × Including/Excluding).
- [x] Strengthened `test_cve_lookup_empty_results` to assert Stage 2 multi-call behavior.
- [x] No real network calls in tests.
- [x] No new API keys or third-party services required.
- [x] New dependency (`packaging`) added to `requirements.txt`, `pyproject.toml`, and `uv.lock` regenerated.
- [x] Fail-open behavior for unparseable target versions (return all Stage 2 CVEs rather than dropping them).
- [x] Defensive handling of unparseable constraint versions in cpeMatch (skip, don't crash).
- [x] Defensive `or []` for null-safety on `configurations`/`nodes`/`cpeMatch` iteration.
- [x] Stage 1 errors do NOT fall through to Stage 2 (verified by test).
- [x] Commit message references `#105`.

## Related

- Issue: #105
- Affects: `tools/cve_tool.py`, `tests/test_cve_lookup.py`. No downstream `full_recon` signal consumer for CVE (verified).

### CODE REVIEW

**Review by gaoharimran29-glitch — CHANGES_REQUESTED** [review]
@01luyicheng  Thanks for the contribution! I tested the implementation against several real NVD API responses, and I really like the overall multi-stage lookup approach. The structure of the implementation is clean and follows the proposed workflow from the issue.

However, while testing, I found a few cases where the current version filtering logic can still produce incorrect results. I'd like to request some changes before merging.

1. Stage 1 results are not version-aware

* The current implementation immediately returns all results from `keywordSearch="<software> <version>"` without validating whether the supplied version is actually affected.
* For example, when testing `OpenSSH 9.8`, the tool returned `CVE-2024-39894`, even though the CVE explicitly states that versions `9.5 through 9.7 before 9.8` are affected. Version `9.8` itself is not vulnerable.
* Since NVD's `keywordSearch` is text-based, it can still return false positives even when the supplied version is not affected.
* I would recommend applying version filtering whenever configuration/version constraint data is available, regardless of whether the result came from Stage 1 or Stage 2.

2. The `criteria` field is not validated

* During the fallback search (`keywordSearch=software`), the implementation only checks whether the supplied version falls within the version constraints.
* It does not verify whether the matched CPE actually corresponds to the requested software.
* Broad searches such as `keywordSearch=nginx` can return vulnerabilities for related or entirely different products/platforms.
* Validating the product portion of the CPE (`criteria`) before performing version comparisons would help avoid incorrect matches.

3. Nested configuration nodes are not handled

* `_cve_affects_version()` currently iterates only through:

  * `configurations -> nodes -> cpeMatch`
* However, the NVD schema also supports nested `children` nodes with logical `AND`/`OR` operators.
* Vulnerabilities that use nested configuration structures may therefore be missed during version matching.
* The filtering logic should recursively handle nested configuration nodes.

4. Vulnerable CPE entries without version constraints are treated as affecting every version

* `_version_in_range()` currently returns `True` whenever none of the following fields are present:

  * `versionStartIncluding`
  * `versionStartExcluding`
  * `versionEndIncluding`
  * `versionEndExcluding`
* While testing `nginx 1.24.0`, I found that `CVE-2009-2629` was returned even though the nginx-specific version ranges only affect versions up to `0.8.x`.
* The reason is that the CVE also contains vulnerable Debian Linux CPE entries without any version constraints. Since `_version_in_range()` returns `True` when no constraints are present and the `criteria` field is not validated, the supplied nginx version is incorrectly treated as vulnerable.
* If version constraints are not present, we cannot reliably determine whether the supplied version is affected, so these cases should not automatically evaluate to `True`.

5. Invalid version handling

* If the supplied version cannot be parsed by `packaging.version.Version`, `_cve_affects_version()` currently returns `True`.
* This effectively causes all matching software CVEs to be returned during the fallback search.
* Returning every CVE for an invalid version string is probably not the intended behaviour.
* It would be better to either return an error or handle invalid versions explicitly before applying version-aware filtering.

Overall, the implementation is very close to the proposed solution, but the current version filtering logic can still introduce both false positives and false negatives in several scenarios. Addressing the points above would make the tool significantly more accurate and align better with the goal of providing reliable version-aware CVE matching.

File: tools/cve_tool.py [hunk 1]
```diff
@@ -1,9 +1,93 @@
...
+    """Query the NVD CVE API by keyword and return the raw vulnerabilities list."""
+    response = requests.get(
+        "https://services.nvd.nist.gov/rest/json/cves/2.0",
+        params={"keywordSearch": keyword},
+        timeout=60,
+        headers={"User-Agent": "CyberSecurity-MCP-Server/1.0"},
+    )
+    response.raise_for_status()
+    data = response.json()
+    return data.get("vulnerabilities", [])
```
[the remark this answers is not in the document]
Comment (01luyicheng) [reply tools/cve_tool.py:23]: Good point about pagination - one clarification: the NVD API defaults to `resultsPerPage=2000` per response (not 20), which covers all but the most extreme cases (e.g. a bare product name with thousands of historical CVEs). I considered adding auto-pagination but decided against it for now because the NVD public API rate-limits to 5 requests / 30s without an API key, and auto-pagination could easily trip that limit. I've documented this rationale in `_query_nvd`'s docstring (commit 2bbb8f4). If a maintainer wants pagination added later (ideally with an API key) I'm happy to follow up.

File: tools/cve_tool.py [hunk 2]
```diff
@@ -1,9 +1,93 @@
...
+        # If we can't parse the target version, don't filter (include the CVE).
+        return True
+    for configuration in cve.get("configurations") or []:
+        for node in configuration.get("nodes") or []:
+            for match in node.get("cpeMatch") or []:
+                if not match.get("vulnerable", False):
+                    continue  # Skip non-vulnerable CPE matches
+                if _version_in_range(target, match):
+                    return True
+    return False
```
[the remark this answers is not in the document]
Comment (01luyicheng) [reply tools/cve_tool.py:127]: Fixed in commit 2bbb8f4. `_cve_affects_version` now traverses `children[]` recursively via an inner `_node_matches` helper, so CVEs that use AND/OR compositions with nested children are matched correctly. Covered by `test_nested_children_nodes_are_traversed` and `test_nested_children_with_non_matching_criteria_excluded`.

File: tools/cve_tool.py [hunk 3]
```diff
@@ -12,34 +96,30 @@ def cve_lookup(software: str, version: str) -> dict:
...
+                "version": version,
+                "total_results": len(items),
+                "cves": [_simplify_cve(item) for item in items],
+                "version_filtering_applied": False,
+            }
+        # Stage 2 & 3: broader query + version filtering
+        items = _query_nvd(software)
+        filtered = [
+            item for item in items
+            if _cve_affects_version(item.get("cve", {}), version)
```
[the remark this answers is not in the document]
Comment (01luyicheng) [reply tools/cve_tool.py]: Fixed in commit 2bbb8f4. The fail-open path is gone - unparseable target versions now return an explicit error before any NVD call (`cve_lookup('x', 'abc')` -> `{'success': False, 'error': 'Invalid version format: ...'}`). As a result, `version_filtering_applied` is `True` on every successful response, so the flag is no longer misleading. Covered by `test_invalid_version_returns_error_without_api_call`.

File: tools/cve_tool.py [hunk 4]
```diff
@@ -1,9 +1,93 @@
...
     """
     Look up known CVEs for a software name and version using the NVD API.
+
+    Uses a 3-stage strategy to reduce false negatives:
+      - Stage 1: keyword search with "<software> <version>".
+      - Stage 2: fallback keyword search with just "<software>".
+      - Stage 3: filter Stage 2 CVEs by parsing cpeMatch version ranges.
+
+    CVEs without structured ``configurations`` are excluded from Stage 2/3
+    results because their version impact cannot be confirmed.
```
[the remark this answers is not in the document]
Comment (01luyicheng) [reply tools/cve_tool.py]: Fixed in commit 2bbb8f4. The docstrings on `_cve_affects_version` and `cve_lookup` have been rewritten to reflect the new contract: filtering is applied to both Stage 1 and Stage 2, CVEs without structured `configurations` are excluded, and the fail-open path for unparseable target versions has been replaced by an upfront error in `cve_lookup`.

File: tests/test_cve_lookup.py [hunk 5]
```diff
@@ -95,5 +125,303 @@ def test_cve_lookup_exception_caught(self, _):
...
+        }
+
+        mock_get.side_effect = [empty_response, data_response]
+
+        result = cve_lookup("someproduct", "abc")
+
+        self.assertTrue(result["success"])
+        self.assertEqual(result["version_filtering_applied"], True)
+        # Unparseable target version => filter is skipped, all CVEs included.
+        self.assertEqual(len(result["cves"]), 2)
```
[the remark this answers is not in the document]
Comment (01luyicheng) [reply tests/test_cve_lookup.py]: Fixed in commit 2bbb8f4. The old `test_invalid_target_version_skips_filter` (which asserted `version_filtering_applied == True` while filtering was skipped) is replaced by `test_invalid_version_returns_error_without_api_call`, which asserts that an unparseable version returns an explicit `success: False` error without making any NVD call. The misleading contract is gone.

**Comment by 01luyicheng** [comment]
Thanks @gaoharimran29-glitch for the detailed review and for testing against real NVD responses - really appreciated. I've addressed all 5 points in commit 2bbb8f4; 28 tests passing (was 20).

- [x] **1. Stage 1 results are now version-aware.** Stage 1 results are passed through `_cve_affects_version`; if all are filtered out, the tool falls back to Stage 2. Fixes the OpenSSH 9.8 / CVE-2024-39894 false positive you mentioned.
- [x] **2. CPE `criteria` field is now validated.** Added `_criteria_matches_software(criteria, software)` which parses the CPE 2.3 (and 2.2) string and matches the requested software against the vendor or product field (case-insensitive, `*` wildcard). Broad nginx searches no longer return Debian/OS-level CVEs.
- [x] **3. Nested `children[]` nodes are handled recursively.** `_cve_affects_version` now traverses nested children via an inner `_node_matches` helper, so CVEs that use AND/OR compositions are matched correctly.
- [x] **4. Constraint-less vulnerable cpeMatch entries no longer evaluate to True.** `_version_in_range` returns `False` when none of the four boundary fields are present, since we can't reliably confirm impact. Fixes the nginx 1.24.0 / CVE-2009-2629 false positive.
- [x] **5. Invalid versions return an explicit error.** `Version(version)` is validated upfront in `cve_lookup`; unparseable versions return `{'success': False, 'error': "Invalid version format: ..."}` before any NVD call. `version_filtering_applied` is therefore `True` on every successful response (the misleading fail-open flag is gone).

I also replied to the Copilot inline comments - the pagination note was based on a wrong premise (NVD defaults to `resultsPerPage=2000`, not 20); I documented the rationale for not adding auto-pagination (NVD's 5 req/30s public rate limit) in `_query_nvd`'s docstring.

New tests added for criteria validation (vendor/product/wildcard), nested children traversal, the nginx false-positive scenario, no-constraint exclusion, and the invalid-version error path. The nginx scenario is covered end-to-end by `test_nginx_false_positive_scenario_filtered_out`.

Could you take another look when you have time? Happy to iterate further.

**Review by gaoharimran29-glitch — APPROVED** [review]
@01luyicheng Thanks for addressing the review comments and iterating on the implementation. I re-tested the changes, and the version-aware filtering logic has been significantly improved.

The updated implementation now applies version filtering consistently, validates CPE criteria, handles nested configuration nodes, properly excludes unconstrained version matches, and handles invalid version inputs gracefully. These changes align well with the goals of the original issue and substantially improve the accuracy of CVE matching.

Everything requested in the review has been addressed. Great work on incorporating the feedback.

LGTM. Merging this PR.
